libs/remote_storage/src/s3_bucket.rs

```diff
@@ -83,10 +87,23 @@ impl S3Bucket {
             .or_else("imds", ImdsCredentialsProvider::builder().build())
         };
 
+        // AWS SDK requires us to specify how the RetryConfig should sleep when it wants to back off
+        let sleep_impl: Arc<dyn AsyncSleep> = Arc::new(TokioSleep::new());
+
+        // We do our own retries (see [`backoff::retry`]).  However, for the AWS SDK to enable rate limiting in response to throttling
+        // responses (e.g. 429 on too many ListObjectsv2 requests), we must provide a retry config.  We set it to use at most one
+        // attempt, and enable 'Adaptive' mode, which causes rate limiting to be enabled.
+        let mut retry_config = RetryConfigBuilder::new();
+        retry_config
+            .set_max_attempts(Some(1))
+            .set_mode(Some(RetryMode::Adaptive));
+
         let mut config_builder = Config::builder()
             .region(region)
             .credentials_cache(CredentialsCache::lazy())
-            .credentials_provider(credentials_provider);
+            .credentials_provider(credentials_provider)
+            .sleep_impl(SharedAsyncSleep::from(sleep_impl))
```

Comment:
what about cancellation for this sleep implementation?

Reply:
S3Bucket already doesn't have cancellation: we'll get stuck in here the same way we would already get stuck if the remote server was slow to respond.  Building proper cancellation into S3Bucket is a todo list item.

The tradeoff is that "cancellability" gets slightly worse compared with backoff::retry, and in return we will have less of a problem with SlowDown storms.  Then eventually we should fix S3Bucket to respect a cancellation token anyway, by extending this sleep_impl.

Reply:
The [`AsyncSleep`](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/config/trait.AsyncSleep.html) trait's `sleep` function seems to not return a result :)

Reply:
Right, but all we need it to do is return when cancelled so that it doesn't wait around.  Then we still have the problem of the actual request lifecycle not being cancellable, but that's our existing problem (unrelated to SlowDown handling)

Reply:
> Right, but all we need it to do is return when cancelled so that it doesn't wait around

fair, good point.